zcash_client_sqlite/src/wallet.rs

```diff
@@ -2292,13 +2367,111 @@ pub(crate) fn put_tx_data(
         ":expiry_height": u32::from(tx.expiry_height()),
         ":raw": raw_tx,
         ":fee": fee.map(u64::from),
+        ":target_height": target_height.map(u32::from),
     ];
 
     stmt_upsert_tx_data
-        .query_row(tx_params, |row| row.get::<_, i64>(0))
+        .query_row(tx_params, |row| row.get::<_, i64>(0).map(TxRef))
         .map_err(SqliteClientError::from)
 }
 
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub(crate) enum TxQueryType {
+    Status,
+    Enhancement,
+}
+
+impl TxQueryType {
+    pub(crate) fn code(&self) -> i64 {
+        match self {
+            TxQueryType::Status => 0,
+            TxQueryType::Enhancement => 1,
+        }
+    }
+
+    pub(crate) fn from_code(code: i64) -> Option<Self> {
+        match code {
+            0 => Some(TxQueryType::Status),
+            1 => Some(TxQueryType::Enhancement),
+            _ => None,
+        }
+    }
+}
+
+pub(crate) fn queue_tx_retrieval(
+    conn: &rusqlite::Transaction<'_>,
+    txids: impl Iterator<Item = TxId>,
+    query_type: TxQueryType,
+    dependent_tx_ref: Option<TxRef>,
+) -> Result<(), SqliteClientError> {
+    // Add an entry to the transaction retrieval queue if we don't already have raw transaction data.
+    let mut stmt_insert_tx = conn.prepare_cached(
+        "INSERT INTO tx_retrieval_queue (txid, query_type, dependent_transaction_id)
+        SELECT :txid, :query_type, :dependent_transaction_id
+        -- do not queue enhancement requests if we already have the raw transaction
```

Comment:
Is it possible that we have the raw transaction but don't have its mined height?

Reply:
I think it is possible, because `wallet::put_tx_data` sets `raw` but not `mined_height`. Therefore this optimization is incorrect given that a Status request can be coalesced with an Enhancement request.

Reply:
Oh, I think it's fine. There are two nontrivial cases:
* If this is an Enhancement request and we have the raw tx, but there is an existing Status request for this txid, then the WHERE clause will be false and the SELECT won't match, so there will be no conflict and we will correctly leave the queue as-is.
* In any case where we already have an Enhancement request, this statement can't affect it. Even if it were possible to have an outstanding Enhancement request for a tx where `raw` is already set, it would not prevent the mined status from being updated by that request.